### PR TITLE
FIX-#3755: return positional indices for '.indices' GroupBy property

### LINKED ISSUE #3755: Modin GroupBy object returns the same values for `.indices` and `.groups` when it shouldn't
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): f79cb85c64476b65ef482c0dd576ef0f258171b8
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd
from modin.utils import try_cast_to_pandas

df = pd.DataFrame({"a": [1, 1, 2, 2]}, index=["a", "b", "c", "d"])

md_grp = df.groupby("a")

print("Modin:\n")
print(f"indices:\n{md_grp.indices}")  # Prints index labels of the groups
print(f"groups:\n{md_grp.groups}\n")  # Prints index label of the groups

pd_grp = try_cast_to_pandas(df).groupby("a")

print("Pandas:\n")
print(f"indices:\n{pd_grp.indices}")  # Prints index labels of the groups
print(f"groups:\n{pd_grp.groups}")  # Prints index label of the groups
```

Output:
```
Modin:

indices:
{1: ['a', 'b'], 2: ['c', 'd']}
groups:
{1: ['a', 'b'], 2: ['c', 'd']}

Pandas:

indices:
{1: array([0, 1]), 2: array([2, 3])}
groups:
{1: ['a', 'b'], 2: ['c', 'd']}
```

### Describe the problem
Going by pandas docs, [`groupby.indices`](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.indices.html#pandas.core.groupby.GroupBy.indices) returns a map of {group name -> group positional indices} and the [`groupby.groups`](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.groups.html#pandas.core.groupby.GroupBy.groups) returns {group name -> group index labels}, however Modin returns group index labels for both of the properties.

The reason for that is that both of the properties return the same value stored in the `._index_grouped`, that holds a value for `.groups` property:
https://github.com/modin-project/modin/blob/f79cb85c64476b65ef482c0dd576ef0f258171b8/modin/pandas/groupby.py#L259-L261
https://github.com/modin-project/modin/blob/f79cb85c64476b65ef482c0dd576ef0f258171b8/modin/pandas/groupby.py#L167-L169

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

Fixes implementation of the [`GroupBy.indices`](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.indices.html#pandas.core.groupby.GroupBy.indices). Now it properly returns <i>positional indices</i> of the groups, instead of labels-based ones as it was done before. The actual method that computes `groups` and `indices` was changed as follows: it was remade from a property (`._index_grouped`) to a method (`._compute_index_grouped()`) taking a boolean flag, indicating whether to compute positional or labels-based group indices.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3755 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/groupby.py [hunk 1]
```diff
@@ -806,89 +822,94 @@ def _iter(self):
...
-                    by = self._by
-                if self._axis == 0:
-                    self._index_grouped_cache = self._index.groupby(by)
-                else:
-                    self._index_grouped_cache = self._columns.groupby(by)
-        return self._index_grouped_cache
+                by = try_cast_to_pandas(by, squeeze=True)
+                pandas_df = self._df._to_pandas()
+            by = wrap_into_list(by, level)
+            groupby_obj = pandas_df.groupby(by=by)
```
Comment (vnlitvinov) [inline modin/pandas/groupby.py:890]: this feels really weird, that we're using pandas here directly (i.e. not at the low-level after query compiler did its work, but right in the frontend), maybe we should make an issue to rectify this?..

File: modin/pandas/test/utils.py [hunk 2]
```diff
@@ -1349,3 +1351,14 @@ def _make_default_file(filename=None, nrows=NROWS, ncols=2, force=True, **kwargs
...
+
+def dict_equals(dict1, dict2):
+    """Check whether two dictionaries are equal and raise an ``AssertionError`` if they aren't."""
+    dict1 = OrderedDict(sorted(dict1.items(), key=lambda item: item[0]))
+    dict2 = OrderedDict(sorted(dict2.items(), key=lambda item: item[0]))
+    for (key1, value1), (key2, value2) in itertools.zip_longest(
+        dict1.items(), dict2.items()
+    ):
+        assert (key1 == key2) or (np.isnan(key1) and np.isnan(key2))
+        np.testing.assert_array_equal(value1, value2)
```
Comment (vnlitvinov) [inline modin/pandas/test/utils.py:1364]:
```suggestion
    for key1, key2 in itertools.zip_longest(sorted(dict1), sorted(dict2)):
        assert (key1 == key2) or (np.isnan(key1) and np.isnan(key2))
        np.testing.assert_array_equal(dict1[key1], dict2[key2])
```
I think it's easier to read

File: modin/pandas/groupby.py [hunk 3]
```diff
@@ -164,9 +163,17 @@ def __bytes__(self):
...
 
+    _groups_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
     @property
     def groups(self):
-        return self._index_grouped
+        if self._groups_cache is not no_default:
+            return copy.deepcopy(self._groups_cache)
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/pandas/groupby.py]: Okay, just found out that `.indices` and `.groups` are properties and it's okay for them to be changed by users from outside:
```python
>>> grp = pandas.DataFrame({"a": [1, 1, 2, 2]}).groupby("a")
>>> indices = grp.indices
>>> grp.indices # Original indices
{1: array([0, 1], dtype=int64), 2: array([2, 3], dtype=int64)}
>>> indices["new_key"] = "new_value"
>>> indices[1][0] = "random_value"
>>> indices[1][0] = 12345
>>> grp.indices # Mutated indices
{1: array([12345, 1], dtype=int64), 2: array([2, 3], dtype=int64), 'new_key': 'new_value'}
```

Removed this copying.

File: modin/pandas/test/utils.py (hunk 2 above)
Comment (dchigarev) [reply modin/pandas/test/utils.py]: applied

File: modin/pandas/groupby.py (hunk 1 above)
Comment (dchigarev) [reply modin/pandas/groupby.py:889]: Created an issue for that: #3771

**Review by vnlitvinov — DISMISSED** [review]
LGTM!

**Review by vnlitvinov — APPROVED** [review]

**Review by devin-petersohn — APPROVED** [review]
Thanks @dchigarev this LGTM!

**Pull request merged**